src/Kestrel.Core/Internal/Http2/Http2FrameReader.cs

```diff
@@ -3,40 +3,245 @@
 
 using System;
 using System.Buffers;
+using System.Buffers.Binary;
+using System.Collections.Generic;
+using System.Diagnostics;
 
 namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2
 {
-    public static class Http2FrameReader
+    public class Http2FrameReader
     {
-        public static bool ReadFrame(ReadOnlySequence<byte> readableBuffer, Http2Frame frame, uint maxFrameSize, out SequencePosition consumed, out SequencePosition examined)
+        /* https://tools.ietf.org/html/rfc7540#section-4.1
+            +-----------------------------------------------+
+            |                 Length (24)                   |
+            +---------------+---------------+---------------+
+            |   Type (8)    |   Flags (8)   |
+            +-+-------------+---------------+-------------------------------+
+            |R|                 Stream Identifier (31)                      |
+            +=+=============================================================+
+            |                   Frame Payload (0...)                      ...
+            +---------------------------------------------------------------+
+        */
+        public const int HeaderLength = 9;
+
+        private const int LengthOffset = 0;
+        private const int TypeOffset = 3;
+        private const int FlagsOffset = 4;
+        private const int StreamIdOffset = 5;
+
+        public const int SettingSize = 6; // 2 bytes for the id, 4 bytes for the value.
+
+        // No frame type needs more than 8 bytes of additional fields. Payload data beyond the known fields is not stored in this buffer.
+        private readonly byte[] _extendedHeader = new byte[HeaderLength + 8];
```

Comment:
Why not make this a static field instead? That would allow us to keep this class static and I believe static classes have a small performance benefit. Did a quick test on a local branch and it seems to work.

Reply:
Nevermind, we need a buffer per connection.